wandb/sdk/lib/service_connection.py

```diff
@@ -0,0 +1,200 @@
+) -> ServiceConnection:
+    """Connects to the service process, starting one up if necessary."""
+    conn = _try_connect_to_existing_service()
+    if conn:
+        return conn
+
+    return _start_and_connect_service(settings)
+
+
+def _try_connect_to_existing_service() -> ServiceConnection | None:
+    """Attemps to connect to an existing service process."""
+    token = service_token.get_service_token()
+    if not token:
+        return None
+
+    # Only localhost sockets are supported below.
+    assert token.host == "localhost"
+    client = SockClient()
+
+    try:
+        # TODO: This may block indefinitely if the service is unhealthy.
+        client.connect(token.port)
+
+    except Exception as e:
+        raise WandbServiceConnectionError(
+            "Failed to connect to internal service."
+        ) from e
+
+    return ServiceConnection(client=client, proc=None)
+
+
+def _start_and_connect_service(
+    settings: wandb_settings.Settings,
+) -> ServiceConnection:
+    """Starts a service process and returns a connection to it.
+
+    An atexit hook is registered to tear down the service process and wait for
+    it to complete. The hook does not run in processes started using the
+    multiprocessing module.
+    """
+    proc = service._Service(settings)
+    proc.start()
+
+    port = proc.sock_port
+    assert port
+    client = SockClient()
+    client.connect(port)
+
+    service_token.set_service_token(
+        parent_pid=os.getpid(),
+        transport="tcp",
+        host="localhost",
+        port=port,
+    )
+
+    def teardown_atexit():
+        conn.teardown(hooks.exit_code)
+
+    conn = ServiceConnection(
+        client=client,
+        proc=proc,
+        cleanup=lambda: atexit.unregister(teardown_atexit),
+    )
+
+    hooks = ExitHooks()
+    hooks.hook()
+    atexit.register(teardown_atexit)
+
+    return conn
+
+
+class ServiceConnection:
+    """A connection to the W&B internal service process."""
+
+    def __init__(
+        self,
+        client: SockClient,
+        proc: service._Service | None,
+        cleanup: Callable[[], None] | None = None,
+    ):
+        """Returns a new ServiceConnection.
+
+        Args:
+            client: A socket that's connected to the service.
+            proc: The service process if we own it, or None otherwise.
+            cleanup: A callback to run on teardown before doing anything.
+        """
+        self._client = client
+        self._proc = proc
+        self._torn_down = False
+        self._cleanup = cleanup
+
+    def make_interface(self, mailbox: Mailbox) -> InterfaceBase:
+        """Returns an interface for communicating with the service."""
+        return InterfaceSock(self._client, mailbox)
+
+    def send_record(self, record: pb.Record) -> None:
+        """Sends data to the service."""
+        self._client.send_record_publish(record)
+
+    def inform_init(
+        self,
+        settings: wandb_settings_pb2.Settings,
+        run_id: str,
```

Comment:
what do you think about renaming these to stream_id, which is more accurate, technically these don't have to be the same as run_id -  it is just the case currently

Reply:
I agree with the idea, but it's slightly scary in this case because this parameter tends to get passed by name, so I'd have to find and update all callsites.

Reply:
ok, we can do it later on... wish we did it from the start